aip/general/0162.md

```diff
@@ -25,201 +25,170 @@ to the version of an API as a whole.
 
 ## Guidance
 
-APIs **may** store a revision history for a resource if it is useful to users.
+APIs **may** store a revision history for a resource. Examples of when it is
+useful include:
 
-APIs implementing resources with a revision history **must** provide a
-`revision_id` field on the resource:
+- When it is valuable to expose older versions of a resource via an API. This
+  can avoid the overhead of the customers having to write their own API to store
+  and enable retrieval of revisions.
+- Other resources depend on different revisions of a resource.
+- There is a need to visualize the change of a resource over time.
 
-```proto
-message Book {
-  // The name of the book.
-  string name = 1;
+APIs implementing resources with a revision history **should** abstract resource
+revisions as nested collection of the resource. Sometimes, the revisions
+collection can be a top level collection, exceptions include:
+
+- If resource revisions are meant to have longer lifespan than the parent
+resource. In other words, resource revisions exist after the resource deletion.
 
-  // Other fields…
+The resource revision type **must** inherit from the parent resource and has a
+field being the parent resource type.
+
+```proto
+message BookRevision {
+   // The name of the book revision.
+   string name = 1;
 
-  // The revision ID of the book.
-  // A new revision is committed whenever the book is changed in any way.
-  // The format is an 8-character hexadecimal string.
-  string revision_id = 5 [
-    (google.api.field_behavior) = IMMUTABLE,
-    (google.api.field_behavior) = OUTPUT_ONLY];
+   // The snapshot of the book
+   Book book = 2;
 
   // The timestamp that the revision was created.
-  google.protobuf.Timestamp revision_create_time = 6
+  google.protobuf.Timestamp create_time = 3
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+
+  // Other revision IDs that share the same snapshot.
+  repeated string alternate_ids = 4
```

Comment:
So if I call GetBookRevision with a name of "books/xyz/revisions/latest" or a tag, what should happen? Should `name` be the name I requested, and the canonical name is returned in alternate_ids, or is the same representation returned in all cases, with the name being the canonical name, and alternate_ids including "latest" and any tags created? I suspect the latter is more useful, but I don't know whether it *could* confuse Terraform etc.

Reply:
Terraform likely won't use revisions, but definitely good to consider it. 

I think it has to be the former: returning the resource name the user sent. Otherwise, it's deviating behavior specified in the standard method AIPs.

One of my personal goals here is to remove exceptional cases, especially in the standard methods: if we introduce scenarios where a GET of a resource with an ID foo returns back a resource with an ID bar, that's another potential behavior exception that clients must follow. WDYT?

Reply:
It feels to me like the latter is closer to https://google.aip.dev/122#resource-id-aliases:

> APIs may provide programmatic aliases for common lookup patterns. However, all data returned from the API must use the canonical resource name.

Is the "always return what the user asked for" specified anywhere, other than for project name/ID in AIP-2510?

Reply:
given that this is documented in 122, should I add the clause here?

Reply:
I'll leave it in 122 for now (since duplication often leads to drift).